zebra/zapi_msg.c

```diff
@@ -1689,100 +1690,215 @@ static void zread_route_add(ZAPI_HANDLER_ARGS)
+
+	return;
+
+stream_failure:
+	flog_warn(EC_ZEBRA_NEXTHOP_CREATION_FAILED,
+		  "%s: Nexthop group deletion failed", __func__);
+	return;
+}
+
+static void zread_nhg_add(ZAPI_HANDLER_ARGS)
+{
+	struct stream *s;
+	uint32_t id;
+	size_t nhops, i;
+	struct zapi_nexthop zapi_nexthops[MULTIPATH_NUM];
+	struct nexthop_group *nhg = NULL;
+	uint16_t proto;
+	struct nhg_hash_entry *nhe;
+
+	s = msg;
+
+	STREAM_GETW(s, proto);
+	STREAM_GETL(s, id);
+	STREAM_GETW(s, nhops);
+
+	if (zserv_nexthop_num_warn(__func__, NULL, nhops))
+		return;
 
-		bnhg = zebra_nhg_backup_alloc();
-		nexthop = NULL;
-		last_nh = NULL;
+	if (nhops <= 0) {
+		flog_warn(EC_ZEBRA_NEXTHOP_CREATION_FAILED,
+			  "%s: No nexthops sent", __func__);
+		return;
 	}
 
-	/* Copy backup nexthops also, if present */
-	for (i = 0; i < api.backup_nexthop_num; i++) {
-		api_nh = &api.backup_nexthops[i];
+	for (i = 0; i < nhops; i++) {
+		struct zapi_nexthop *znh = &zapi_nexthops[i];
 
-		/* Convert zapi backup nexthop */
-		nexthop = nexthop_from_zapi(re, api_nh, &api);
-		if (!nexthop) {
-			flog_warn(
-				EC_ZEBRA_NEXTHOP_CREATION_FAILED,
-				"%s: Backup Nexthops Specified: %d but we failed to properly create one",
-				__func__, api.backup_nexthop_num);
-			nexthop_group_delete(&ng);
-			zebra_nhg_backup_free(&bnhg);
-			XFREE(MTYPE_RE, re);
+		if (zapi_nexthop_decode(s, znh, 0, 0) != 0) {
+			flog_warn(EC_ZEBRA_NEXTHOP_CREATION_FAILED,
+				  "%s: Nexthop creation failed", __func__);
 			return;
 		}
+	}
 
-		/* Backup nexthops can't have backups; that's not valid. */
-		if (CHECK_FLAG(nexthop->flags, NEXTHOP_FLAG_HAS_BACKUP)) {
-			if (IS_ZEBRA_DEBUG_RECV) {
-				nexthop2str(nexthop, nhbuf, sizeof(nhbuf));
-				zlog_debug("%s: backup nh %s with BACKUP flag!",
-					   __func__, nhbuf);
-			}
-			UNSET_FLAG(nexthop->flags, NEXTHOP_FLAG_HAS_BACKUP);
-			nexthop->backup_num = 0;
-		}
+	if (!zapi_read_nexthops(client, NULL, zapi_nexthops, 0, 0, nhops, 0,
+				&nhg, NULL)) {
+		flog_warn(EC_ZEBRA_NEXTHOP_CREATION_FAILED,
+			  "%s: Nexthop Group Creation failed", __func__);
+		return;
+	}
 
-		if (CHECK_FLAG(api.message, ZAPI_MESSAGE_SRTE)) {
-			SET_FLAG(nexthop->flags, NEXTHOP_FLAG_SRTE);
-			nexthop->srte_color = api_nh->srte_color;
-		}
+	/*
+	 * Create the nhg
+	 */
+	nhe = zebra_nhg_proto_add(id, proto, nhg, 0);
 
-		/* MPLS labels for BGP-LU or Segment Routing */
-		if (CHECK_FLAG(api_nh->flags, ZAPI_NEXTHOP_FLAG_LABEL)
-		    && api_nh->type != NEXTHOP_TYPE_IFINDEX
-		    && api_nh->type != NEXTHOP_TYPE_BLACKHOLE
-		    && api_nh->label_num > 0) {
+	nexthop_group_delete(&nhg);
 
-			label_type = lsp_type_from_re_type(client->proto);
-			nexthop_add_labels(nexthop, label_type,
-					   api_nh->label_num,
-					   &api_nh->labels[0]);
-		}
+	/*
+	 * TODO:
+	 * Assume fully resolved for now and install.
+	 *
+	 * Resolution is going to need some more work.
+	 */
+	if (nhe)
+		nhg_notify(proto, client->instance, id, ZAPI_NHG_INSTALLED);
```

Comment:
mmm - we shouldn't tell the owner/daemon the result until we have the result. we have a results callback that is being used - can the nhg results be moved there, to align with what we do for route redist and nht?

Reply:
sure that makes sense. This was just temp for now since we are only doing onlink and they are expected to always be success.

We will still may need to notify fail here since it can early fail in the add function if it doesnt meet a few params.

Reply:
Hmm.. so this is a bit more complex than I had thought originally. We have to do instance tracking since there is the possibility of having multiple instance of the protos. It will require a lot of code changes in the NHE and dataplane code to address this from the looks of it to get it closer to the way routes track it.

Since this is a fairly basic use case for now. We arent doing resolution or much with these proto nexthops yet, zebra really just sends them off to the dataplane. I'd like to postpone adding that code until next iteration so we can get this patch in. Though I definitely agree it should be moved to consolidate into that path since the way it is right now was just to simplify it.

Reply:
but ... doesn't this contradict what you wrote when I asked about the sharpd behavior above? `zebra_nhg_proto_add()` is going to be able to allocate an nhg/nhe - and this test will cause zebra to always send sharpd an INSTALLED notification. that will cause sharpd to only use the "installed" nhgs - but though the way you've defined the zapi messaging and the apis doesn't include backup nexthops, so sharpd will no longer be able to test backups.

Reply:
it would only mark success if it met the condition that its onlink which should just install success anyway.

Though I do see your concern with backup nexthop objects. I'm not really willing to support those just yet until i figure out how to handle resolution. I think the support will come naturally through that.

Would you be okay with me failing if a backup nexthop has been given for now? So it will always revert to the old way with those then?